Add tutorial (#421)
pester/docs: tutorial/2-testing-a-module/3-public-functions.mdx

---

The test you have works, but it is testing something slightly different from what your users will run. It dot-sources one `.ps1` file directly. Your users import a module. This page closes that gap.

## Why dot-sourcing is not enough

Dot-sourcing `Get-Planet.ps1` loads that one file and nothing else. That means your test suite cannot tell you:

- whether `Get-Planet` is actually **exported** — a function missing from `FunctionsToExport` passes every dot-sourced test and is invisible to users
- whether the function works when its **private helpers** are loaded alongside it, which is the only way it ever runs in reality

Importing the module instead fixes both. The test then calls the function through the same front door your users do, and a broken export becomes a test failure instead of a support ticket.

## Importing the module

Change the `BeforeAll` in your test file:

```powershell title="Planetarium/Public/Get-Planet.Tests.ps1"
BeforeAll {
    # diff-remove
    . $PSCommandPath.Replace('.Tests.ps1', '.ps1')
    # diff-add

>     Import-Module $PSScriptRoot/../Planetarium.psd1 -Force

Author: ```suggestion
    Import-Module "$PSScriptRoot/../Planetarium.psd1" -Force
```

}
```

`$PSScriptRoot` is the folder holding the test file, so `../Planetarium.psd1` walks up from `Public/` to the manifest. Import the **manifest**, not the `.psm1` — the manifest is what your users get, and it is the thing that carries `FunctionsToExport`.

:::tip `-Force` is not optional here
Without it, a module already loaded in your session stays loaded, and you will spend a genuinely upsetting amount of time wondering why your fix changed nothing.
:::

## Filling out the tests

With the module imported, write tests for the behaviour that actually matters:

```powershell title="Planetarium/Public/Get-Planet.Tests.ps1"
BeforeAll {

>     Import-Module $PSScriptRoot/../Planetarium.psd1 -Force

Author: ```suggestion
    Import-Module "$PSScriptRoot/../Planetarium.psd1" -Force
```

Run it:

```powershell
Invoke-Pester -Path ./Planetarium/Public
```

```
Tests Passed: 5, Failed: 0, Skipped: 0, Inconclusive: 0, NotRun: 0
```

## A function with dependencies

Now add a function that leans on the private side of the module. First the two helpers:

```powershell title="Planetarium/Private/ConvertTo-AstronomicalUnit.ps1"
function ConvertTo-AstronomicalUnit {
    param (
        [Parameter(Mandatory)]
        [double] $Kilometre
    )

    [math]::Round($Kilometre / 149597870.7, 3)
}
```

```powershell title="Planetarium/Private/Test-PlanetName.ps1"
function Test-PlanetName {
    param (
        [string] $Name
    )

    $known = 'Mercury', 'Venus', 'Earth', 'Mars', 'Jupiter', 'Saturn', 'Uranus', 'Neptune'
    $Name -in $known
}
```

Then the public function that uses both:

```powershell title="Planetarium/Public/Get-PlanetDistance.ps1"
function Get-PlanetDistance {
    [CmdletBinding()]
    param (
        [Parameter(Mandatory)]
        [string] $Name
    )

    if (-not (Test-PlanetName -Name $Name)) {
        throw "Unknown planet '$Name'."
    }

    $planet = Get-Planet -Name $Name
    ConvertTo-AstronomicalUnit -Kilometre $planet.DistanceFromSunKm
}
```

> This is where importing the module pays off. `Get-PlanetDistance` calls two functions that live in other files. Dot-sourcing `Get-PlanetDistance.ps1` on its own would fail immediately with "The term 'Test-PlanetName' is not recognized". Importing the module loads everything together, exactly as it runs in production.

Author: ```suggestion
This is where importing the module pays off. `Get-PlanetDistance` calls two functions that live in other files. Dot-sourcing `Get-PlanetDistance.ps1` on its own would fail immediately with `The term 'Test-PlanetName' is not recognized`. Importing the module loads everything together, exactly as it runs in production.
```

## Testing it

```powershell title="Planetarium/Public/Get-PlanetDistance.Tests.ps1"
BeforeAll {

>     Import-Module $PSScriptRoot/../Planetarium.psd1 -Force

Author: ```suggestion
    Import-Module "$PSScriptRoot/../Planetarium.psd1" -Force
```

}

Describe 'Get-PlanetDistance' {
    It 'Reports Earth as one astronomical unit from the sun' {
        Get-PlanetDistance -Name 'Earth' | Should-Be 1
    }

    It 'Reports Mercury as closer to the sun than Earth' {
        Get-PlanetDistance -Name 'Mercury' | Should-BeLessThan 1
    }

    It 'Throws for a planet it does not know' {
        { Get-PlanetDistance -Name 'Pluto' } | Should-Throw -ExceptionMessage "Unknown planet 'Pluto'."
    }
}
```

> The error-case test is the interesting one. Note the braces: `{ Get-PlanetDistance -Name 'Pluto' }` is a **script block**, not a call. You are handing Pester the code to run, so it can catch what comes out. Without the braces the exception would escape and fail the test before `Should-Throw` ever saw it.

Author: ```suggestion
The error-case test is the interesting one. Note the braces: `{ Get-PlanetDistance -Name 'Pluto' }` is a **scriptblock**, not a call. You are handing Pester the code to run, so it can catch what comes out. Without the braces the exception would escape and fail the test before `Should-Throw` ever saw it.
```
